### PR TITLE
feat: add MaxTxs, MaxCalldata limit for a block

### PR DESCRIPTION
**Pull request by lightscale-luke**
* Since zkevm-prover have limitations for creating proof, the limitations should be handled in block creating process.
* MaxTxs and MaxCalldata limitations are added.
* Please note that default value of MaxTxs and MaxCalldata can be changed.

### CODE REVIEW

File: core/rawdb/accessors_metadata.go [hunk 1]
```diff
@@ -27,6 +29,11 @@ import (
 	"github.com/ethereum/go-ethereum/rlp"
 )
 
+// ref : https://ethereum.org/en/developers/tutorials/downsizing-contracts-to-fight-the-contract-size-limit/
```
Comment (TomTaehoonKim) [inline core/rawdb/accessors_metadata.go:32]: Let's fix the link to https://eips.ethereum.org/EIPS/eip-170.

File: core/types/block.go [hunk 2]
```diff
@@ -367,6 +367,16 @@ func (b *Block) Size() uint64 {
 	return uint64(c)
 }
 
+// PayloadSize returns the sum of all transactions in a block.
```
Comment (TomTaehoonKim) [inline core/types/block.go:370]: How about fixing the comment to
"// PayloadSize returns the sum of size of all transactions in a block."?

File: core/rawdb/accessors_metadata.go [hunk 3]
```diff
@@ -27,6 +29,11 @@ import (
...
+// 24576 is max calldata for a contract
+const defaultMaxCalldata = 24576 * 5 //122880 = 120 * 1024 in scroll
```
Comment (chokobole) [inline core/rawdb/accessors_metadata.go:34]:
```suggestion
const defaultMaxCalldata = 24576 * 5 // 122880 = 120 * 1024 in scroll
```

File: miner/worker.go [hunk 4]
```diff
@@ -909,6 +910,13 @@ func (w *worker) commitTransactions(env *environment, txs *types.TransactionsByP
 		if tx == nil {
 			break
 		}
+		// [Kroma: START]
```
Comment (chokobole) [inline miner/worker.go:913]: We've marked the features that is derived from `scroll` with this comment!

```suggestion
		// [Scroll: START]
```

File: core/types/block.go (hunk 2 above)
Comment (lightscale-luke) [reply core/types/block.go:370]: will change to 
// PayloadSize returns the sum of all transactions' sizes in a block.

File: cmd/utils/flags.go [hunk 5]
```diff
@@ -2273,8 +2287,18 @@ func MakeChain(ctx *cli.Context, stack *node.Node, readonly bool) (*core.BlockCh
...
+	var overrides core.ChainOverrides
+	if ctx.IsSet(MaxTxsFlag.Name) {
+		overrides.MaxTxs = new(int)
+		*overrides.MaxTxs = ctx.Int(MaxTxsFlag.GetValue())
+	}
+	if ctx.IsSet(MaxCalldataFlag.Name) {
+		overrides.MaxCalldata = new(int)
+		*overrides.MaxCalldata = ctx.Int(MaxCalldataFlag.GetValue())
+	}
+
```
Comment (chokobole) [inline cmd/utils/flags.go:2299]: When following the codes, it seems you don't have to set the flags here. Let me test codes without this and share the result.

File: core/rawdb/accessors_metadata.go [hunk 6]
```diff
@@ -67,6 +67,7 @@ func ReadChainConfig(db ethdb.KeyValueReader, hash common.Hash) *params.ChainCon
 	if config.Kroma != nil {
 		config.Clique = nil // get rid of legacy clique data in chain config (kroma goerli issue)
 	}
+
```
Comment (chokobole) [inline core/rawdb/accessors_metadata.go:70]: Please remove this.

File: params/config.go [hunk 7]
```diff
@@ -697,6 +700,13 @@ func (c *ChainConfig) IsKroma() bool {
 // IsValidTxCount returns whether the given block's transaction count is below the limit.
 func (c *ChainConfig) IsValidTxCount(count int) bool {
 	return c.MaxTxPerBlock == nil || count <= *c.MaxTxPerBlock
+	// return count <= c.MaxTxPerBlock
```
Comment (chokobole) [inline params/config.go:703]: Please remove the comment.

File: params/config.go [hunk 8]
```diff
@@ -697,6 +700,13 @@ func (c *ChainConfig) IsKroma() bool {
...
+}
+
+// IsValidBlockSize returns whether the given block's transaction payload size is below the limit.
+func (c *ChainConfig) IsValidBlockSize(size int) bool {
+	return c.MaxTxPayloadBytesPerBlock == nil || size <= *c.MaxTxPayloadBytesPerBlock
+	// return size <= c.MaxTxPayloadBytesPerBlock
```
Comment (chokobole) [inline params/config.go:709]: Please remove the comment.

File: cmd/utils/flags.go (hunk 5 above)
Comment (chokobole) [reply cmd/utils/flags.go:2299]: When I tested, it isn't needed!

File: cmd/utils/flags.go [hunk 9]
```diff
@@ -672,6 +672,20 @@ var (
...
 		Category:  flags.MiscCategory,
 	}
+	MaxTxsFlag = &cli.IntFlag{
+		Name:     "max.txs",
+		Usage:    "Allowed max number of transactions in a block (default = 200)",
+		Value:    200,
+		Category: flags.MiscCategory,
+	}
+	// https://eips.ethereum.org/EIPS/eip-170.
+	// 24576 is max calldata for a contract, 122880 = 24576 * 5 (120 * 1024 in scroll)
```
Comment (TomTaehoonKim) [inline cmd/utils/flags.go:682]:
```suggestion
	// 24576 is max calldata for a contract, 122880 = 24576 * 5 (same but written 120 * 1024 in scroll)
```

File: core/types/block.go [hunk 10]
```diff
@@ -367,6 +367,16 @@ func (b *Block) Size() uint64 {
 	return uint64(c)
 }
 
+// PayloadSize returns the sum of all transactions' sizes in a block.
+func (b *Block) PayloadSize() int {
+	// add up all txs sizes
```
Comment (TomTaehoonKim) [inline core/types/block.go:372]:
```suggestion
	// add up all txs' sizes
```

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by 0xHansLee — APPROVED** [review]
lgtm

**Review by chokobole — APPROVED** [review]

**Pull request merged**